### PR TITLE
Feat/UI/alert/edit alert screen: EditAlert Screen UI 

### PR DESCRIPTION
**Pull request by agonzalez-r**
# Feat/UI/alert/edit alert screen: EditAlert Screen UI 

## Description
This PR introduces `EditAlert` feature screen. Which should be accessible from any specific alert of `My Alerts` list in `AlertLists`. It allows the user to modify the fields for a specific Alert which passes as parameter to `EditAlertScreen`. 

Since this screen uses the same components than `CreateAlert.kt`, I created a shared components file called `AlertComponents.kt`. The composables shared in that file are: 
- Product Dropdown menu 
- Urgency Dropdown menu 
- Location field and dropdown menu with suggestions
- Message field 
- Validate fields function 
- other private functions shared between components (`ExposedDropdownMenuSample()`)

I took the pre-existing code in `CreateAlert.kt` and modularised it, fixed a few details and made sure that it effectively adapts to both screens (`CreateAlert` and `EditAlert`), while preserving the code as much as possible. I had to make choices in order to prevent too much re-writing the code because that would have taken too long for this task. 

The buttons of each screen are different, therefore they are not in the shared components files, however both screens call `validateFields()` before submitting/saving the Alert. 

**Important**: This is the UI of the screen, I added all the TODOs I could think of for when we implement the functionalities (integrate the Alert ViewModel) as well as things that will need to be removed when everything comes together.

#### Added
- `main/ui/components/AlertComponents.kt`: composables and functions shared by `CreateAlert.kt` and `EditAlert.kt`
- `main/ui/alert/EditAlert.kt`: screen for editing the Alert

#### Modified
- `main/ui/alert/CreateAlert.kt`:adapted to use shared composables from `AlertComponents.kt`

## Testing
For testing, I tested: 
- the correct display of all the components 
- the response of the `Save` button for both successful and unsuccessful inputs (only Location and message could be left empty)
- the navigation of the buttons when the operations are a success

## Screenshots 
Figma design I followed (without the radius selector and organic check box, since these features are not implemented yet): 
Screenshot 2024-11-20 at 13 11 38

Implemented UI: 
Screenshot 2024-11-20 at 11 08 49

### CODE REVIEW

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        Button(
+            modifier = Modifier.wrapContentSize().testTag(RESOLVE_BUTTON),
+            onClick = {
+              // TODO: resolve alert
+              navigationActions.navigateTo(Screen.ALERT_LIST)
+            },
+            colors = getFilledPrimaryContainerButtonColors(),
+        ) {
+          Text(text = RESOLVE_BUTTON_TEXT, style = MaterialTheme.typography.headlineSmall)
+        }
```
Comment (taghizadlaura) [inline app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:186]: Looking at the code for the delete and resolve actions, I noticed that there are no confirmation dialogs in place. This could lead to accidental operations, where a user might delete or resolve an alert unintentionally. For better usability and to prevent these kinds of mistakes, it would be a good idea to add confirmation dialogs for these actions in future iterations. We can discuss this suggestion later, as it's not the primary focus of this PR.

**Review by taghizadlaura — APPROVED** [review]
## Summary
LGTM :
This PR introduces the `EditAlertScreen` composable, enabling users to modify, delete, and resolve alerts. It employs Material 3 components, scaffold-based layouts, and placeholder logic for critical operations. The accompanying test suite thoroughly validates the screen's functionality and behavior, ensuring robustness.

## Important

### Code Quality
LGTM :
- **Readability**: 
The composable layout is well-organized, and the use of `testTags` improves testability and clarity.
Constants like `DELETE_BUTTON_TEXT` and `SAVE_BUTTON_TEXT` centralize text resources and simplify updates.

- **Consistency**:
Consistent application of Material 3 design principles.
Effective use of `TODO` comments for indicating areas requiring further development.

### Functionality
LGTM :
- **Correctness**:
Tests confirm essential operations such as saving, deleting, and resolving alerts work as intended.
`validateFields` provides basic input validation, preventing invalid actions.

- **Edge Cases**:
Tests such as `updateAlertInvalidLocation` and `updateAlertInvalidMessage` validate edge cases.

### Performance
No major performance concerns noted in the implementation or tests.

### Security
Lack of confirmation dialogs for delete and resolve actions poses a usability risk for accidental operations. This should be considered in future iterations to enhance usability and prevent unintentional deletions or resolutions.

## Testing
LGTM :
The test suite is comprehensive, covering:
  1. Proper rendering of UI elements.
  2. Validation of form fields for invalid data.
  3. Behavior of primary actions (`Save`, `Delete`, `Resolve`).
  4. Navigation after successful operations.

Additional suggestions:
  - Simulate more user interactions, such as switching urgency or product, to verify state consistency.

## Documentation
LGTM
The documentation is good overall. 
The `TODO` comments explicitly outline the intended behaviour.

## Suggestions
- Consider adding confirmation dialogs for delete and resolve actions to prevent accidental operations.
- Simulate more user interactions, such as switching urgency or product, to verify state consistency and ensure the UI remains performant.

## Steps before PR approved
I approve your PR, but before merging, please try to fix the SonarCloud issue. I'm not sure why the 'Quality Gate failed,' but it could have consequences in the main branch if we merge it as it is.

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt (hunk 1 above)
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:186]: Yes I will implement `Toast` messages before merging

**Review by lazarinibruno — APPROVED** [review]
## Summary
Thanks for your contribution! Being able to edit the alerts is a step towards reaching our end goal of the app.

## Important

### Code Quality
- Great work creating the shared components. It helps with readability and maintainability. 
- Adding the TODO comment will be useful when integrating the alert view model. Nice addition :)

### Functionality
Tested on my device and the UI works as expected.

## Testing
Thorough tests. They pass locally.

## Documentation
- Good documentation of composables and functions.
- Appropiate comments

## Suggestions
Please read my coments.

## Steps before PR approved
Besides the comments on minor stuff, PR looks great.

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+ *
+ * @param navigationActions Actions to handle navigation events.
+ * @param locationViewModel ViewModel to manage location data.
+ * @param alert The alert object containing the details to be edited.
+ */
+@Composable
+fun EditAlertScreen(
+    navigationActions: NavigationActions,
+    locationViewModel: LocationViewModel,
+    alert: Alert =
```
[the remark this answers is not in the document]
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt:71]: When we integrate this screen, the alert should be passed as parameter (to know which alert to update). This screen will be called from `AlertLists`, `My Alerts` tab. Ideally, when clicked on a certain button (see figma), we call the `EditAlert` screen passing the corresponding alert as parameter.

File: app/src/main/java/com/android/periodpals/ui/components/AlertComponents.kt [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+  // Location Input with dropdown using ExposedDropdownMenuBox
+  ExposedDropdownMenuBox(
+      expanded = showDropdown && locationSuggestions.isNotEmpty(),
+      onExpandedChange = { showDropdown = it }, // Toggle dropdown visibility
+      modifier = Modifier.wrapContentSize(),
+  ) {
+    OutlinedTextField(
+        modifier =
+            Modifier.menuAnchor() // Anchor the dropdown to this text field
```
[the remark this answers is not in the document]
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/components/AlertComponents.kt:126]: As I said on the description of the PR, I just took the previous code form `CreateAlert` and modularised it in order to share it with `EditAlert`, since going over all the code was not the objective of this task.

File: app/src/main/java/com/android/periodpals/ui/components/AlertComponents.kt [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+ * @param testTag The test tag for the dropdown menu.
+ */
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+private fun ExposedDropdownMenuSample(
+    itemsList: List<String>,
+    label: String,
+    defaultValue: String,
+    setIsSelected: (Boolean) -> Unit,
+    onValueChange: (String) -> Unit, // fill Alert values
```
[the remark this answers is not in the document]
Comment (agonzalez-r) [reply app/src/main/java/com/android/periodpals/ui/components/AlertComponents.kt:256]: It's just to facilitate the integration of the Alert VM (this task is done by me) by putting comments where things need to be implemented in the future, I'll make sure to get rid of all these comments once I do it !

**Comment by agonzalez-r** [comment]
About the SonarCloud Quality Gate Fail: 
We increased the Duplicated code limit to 5%, since the group collectively decided that in this case, there was no real issue of duplicated code and therefore the less than 3% requirement was too low. 
So now the PR passes the Quality Gate!
